Update WDPA processing

brynpickering opened the pull request:
* Add WDPA-date config param
* Add schema
* Handle latest batch of WDPA file names (incl `Public` in string) and structures (three separate shapes that need merging)

Commit:
Added wdpa-date config & new WDPA processing

brynpickering commented:
At this point we probably have to accept that with the protected areas data, the process can never be reproducible (we can't store and re-distribute a static dataset, and they update their dataset every few months), so this PR is really just to provide a smoother method to keep up-to-date.

To be even more clever, we could scrape the XML of the online datastore: "http://d1gam3xoknrgr2.cloudfront.net/" to get the URL to the latest dataset. But this is a bit of a hassle...

brynpickering commented:
code snippet on how XML scraping might work to give available date options:

```python
import requests
import xmltodict
import re

search_string = "current/WDPA_([\w]+)_Public_shp.zip"
r = requests.get("http://d1gam3xoknrgr2.cloudfront.net/")
doc = xmltodict.parse(r.content)
contents = doc['ListBucketResult']['Contents']
date_options = set()
for content in contents:
    search_result = re.search(search_string, content["Key"])
    if search_result is not None:
         date_options.update(search_result.groups())
```

This would currently return `{'Mar2021'}`

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile1.zip -d build/raw-wdpa/WDPA_to_merge_0
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile2.zip -d build/raw-wdpa/WDPA_to_merge_0
+        ogrmerge.py -single -o {output.polygons} build/raw-wdpa/WDPA_to_merge_**/*-polygons.shp
+        ogrmerge.py -single -o {output.polygons} build/raw-wdpa/WDPA_to_merge_**/*-points.shp
```
This must be {output.points}.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile2.zip -d build/raw-wdpa/WDPA_to_merge_0
+        ogrmerge.py -single -o {output.polygons} build/raw-wdpa/WDPA_to_merge_**/*-polygons.shp
+        ogrmerge.py -single -o {output.polygons} build/raw-wdpa/WDPA_to_merge_**/*-points.shp
+        exitcode=$?
```
You are only catching the exit code of the second merge.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
-    shell: "curl -sLo {output} -H 'Referer: {URL_PROTECTED_AREAS}' {URL_PROTECTED_AREAS}"
+    params:
+        url = URL_PROTECTED_AREAS.format(config["parameters"]["wdpa-date"])
+    shell: "curl -sLo {output} -H 'Referer: {params.url}' {params.url}"
```
Nice change!

timtroendle commented on config/schema.yaml:
```diff
@@ -8,4 +8,8 @@ properties:
+                description: Indicates the reference NUTS year
+            wdpa-date:
+                type: string
+                enum: [Mar2021, Apr2021]
```
I would not restrict the value in this way as this will make the schema invalid over time. Users will have to touch two places to update. I'd make this a string type instead.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
-        points = "build/raw-wdpa-feb2019/WDPA_Feb2019-shapefile-points.shp"
-    shell: "unzip -o {input} -d build/raw-wdpa-feb2019"
+        polygons = "build/raw-wdpa/wdpa-shapes.shp",
+        polygon_data = "build/raw-wdpa/wdpa-shapes.dbf",
```
This is not generated currently.

timtroendle replied:
@brynpickering  There is an unresolved comment here.

brynpickering replied:
ogrmerge generates the `.dbf` when it generates the associated `.shp`. I suppose I should go through it to see that it includes the expected columns of data (same for the `.shp`).

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
     message: "Extract protected areas data as zip."
     input: rules.raw_protected_areas_zipped.output
+    params:
+        year = config["parameters"]["wdpa-date"]
```
Rather `version` than `year`.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
+    shell:
+        """
+        set +e
+        unzip -o {input} -d build/raw-wdpa
```
They zip zip-files? 🤔

brynpickering replied:
yep...

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
+        """
+        set +e
+        unzip -o {input} -d build/raw-wdpa
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile0.zip -d build/raw-wdpa/WDPA_to_merge_0
```
Can you run a `rm -r build/raw-wdpa/WDPA_to_merge_0` in the end, to delete the temporary files?

If shapefile wasn't such an ugly format, you could the easier and cleaner shadow directive of Snakemake. Unfortunately, it is.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -135,17 +135,38 @@ rule raw_land_cover:
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile0.zip -d build/raw-wdpa/WDPA_to_merge_0
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile1.zip -d build/raw-wdpa/WDPA_to_merge_0
+        unzip -o build/raw-wdpa/WDPA_{params.year}-shapefile2.zip -d build/raw-wdpa/WDPA_to_merge_0
+        ogrmerge.py -single -o {output.polygons} build/raw-wdpa/WDPA_to_merge_**/*-polygons.shp
```
Why the ** in `WDPA_to_merge_**`? The only directory that exists is `WDPA_to_merge_0`, no? If I'm correct, why not renaming that to `WDPA_to_merge`.

timtroendle reviewed (changes requested):
Looks good, but needs a few more changes. Unfortunately, this dataset is really annoying. Thanks for the update.

Commit:
Fixes all over the place

Commit:
Remove temporary WDPA files

brynpickering commented:
Sorry, Friday afternoon fever - lots of errors left over in this implementation that I should have spotted sooner (you know, buy doing some simple checks like running it locally...)! 

I'll make sure it works locally and then let you know when it's ready to look at again :)

timtroendle commented:
#11 anyone?

Commit:
Update WDPA extraction and merging

Commit:
Suppress all errors with ogrmerge

brynpickering commented:
@timtroendle This is now fixed up and works locally. 

The one thing that is a bit annoying is the error catching: merging the files produces thousands of warnings and occasional at the moment (which don't obviously affect information in columns of interest to us). These then stop snakemake in its tracks (it's in 'bash strict mode'). So, I switch off error catching with `ogrmerge` and rely on the fact that if there *is* a problem it will be in the form of the outputs not being successfully generated. Do you think this is the cleanest way of doing it?

timtroendle commented:
Hmm I am not sure to understand. Are you saying that ogrmerge throws errors and that leads Snakemake to crash? That would not be good.

brynpickering commented:
It does throw errors, but not ones that are easy to debug, the output looks like this:

```
Warning 1: Value 555510160 of field WDPAID of feature 115190 not successfully written. Possibly due to too larger number with respect to field width
Warning 1: Value 555510162 of field WDPAID of feature 115191 not successfully written. Possibly due to too larger number with respect to field width
```
+ 998 more lines of much the same
Then:
```
More than 1000 errors or warnings have been reported. No more will be reported from now.
```
The workflow continues unhindered, however, so whatever issue ogrmerge is having doesn't seem to be a problem. However, I could imagine updating this to run in python, such that geopandas handles the merging. Then it would be easier to keep the relevant columns and to debug any issues?

timtroendle commented:
> However, I could imagine updating this to run in python, such that geopandas handles the merging.

I'd feel more comfortable that way. I do not really understand what's going on here. The merging only, without anything else, is a three-liner or so. Plus, it would allow us to move from shape files to geopackage.

Commit:
Merge branch 'main' into update-wdpa-processing

Commit:
Process points and polys at the same time

brynpickering commented:
@timtroendle I've updated this to process the shapefiles purely in python. I've pythonified a bunch of other steps too, which takes advantage of needing to open these shapefiles anyway and having the functionality to clean and limit the scope of the GeoDataFrame using methods in `administrative_borders.py`.

Commit:
Remove excessive whitespace & unnecessary imports

Commit:
Merge branch 'main' into update-wdpa-processing

Commit:
Update WDPA processing to match current main

- Use snakemake 'scripts' to run python script
- Rely on renewablepotentialslib for shape helper functions

Commit:
Add new util tests

brynpickering commented:
@timtroendle is there anything holding this up, or shall I merge?

Commit:
Merge branch 'main' into update-wdpa-processing

Commit:
Update schema

timtroendle commented on config/schema.yaml:
```diff
@@ -8,6 +8,9 @@ properties:
                 enum: [2006, 2010, 2013, 2016, 2021]
                 description: Indicates the reference NUTS year
+            wdpa-version:
+                type: string
```
This could check for the form `MMMYYYY`, but maybe that's just too restrictive.

brynpickering replied:
yeah, who knows how they'll choose to do versioning in future...

timtroendle commented on lib/renewablepotentialslib/shape_utils.py:
```diff
@@ -253,3 +256,27 @@ def study_area(scope_config):
+    original_crs = points.crs
+    # convert points to circles
+    points_in_metres = points.to_crs("epsg:3035")
+    points_in_metres.geometry = [
```
Why not call `points_in_metres.buffer(...)`?

timtroendle replied:
https://geopandas.org/docs/reference/api/geopandas.GeoSeries.buffer.html?highlight=buffer

timtroendle commented on lib/tests/test_shape_utils.py:
```diff
@@ -1,17 +1,23 @@
-    update_features
+    update_features,
+    estimate_polygons_from_points,
+    _radius_meter
```
Is it necessary to test this private function?

brynpickering replied:
I don't see why testing private functions is any less necessary than public ones. Maybe this one in particular doesn't *need* testing, but there's no harm in it, right?

timtroendle commented on lib/tests/test_shape_utils.py:
```diff
@@ -140,3 +146,47 @@ def test_drop_parts_of_geoms(self, capsys, europe_gdf):
+        }
+        point_gdf = gpd.GeoDataFrame(
+            # x = longitude, y = latitude
+            geometry=gpd.points_from_xy([i[1] for i in points.values()], [i[0] for i in points.values()]),
```
This is `geometry=gpd.points_from_xy(zip(*points.values()))`,. no?

brynpickering replied:
It isn't quite, because coordinates are usually communicated as `[lat, lon]`, but `[x, y]` is equivalent to `[lon, lat]`. Anyway, I'll update to be less verbose

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -128,14 +128,32 @@ rule raw_protected_areas_zipped:
+    params:
+        version = config["parameters"]["wdpa-version"]
+    output: temp("build/raw-wdpa")
+    conda: "../envs/default.yaml"
```
This conda env is useless here.

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+def wdpa(path_to_shape_dir, path_to_output, shapes_to_include, scope):
+    all_shapes = gpd.GeoSeries()
+    list_of_shapes_to_include = shapes_to_include.split(",")
+    if not isinstance(list_of_shapes_to_include, list):
```
This can be removed. `split` always returns a list.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -128,14 +128,32 @@ rule raw_protected_areas_zipped:
+        src = script_dir + "protected_areas_in_europe.py",
+        shapes = rules.raw_protected_areas_in_europe_unzipped.output[0]
+    params:
+        shapes_to_include = "polygons,points"
```
Don't encode this list in a string; make it a list!

timtroendle replied:
Also, if you move this to config, the type check can happen in the schema, not within the script.

brynpickering replied:
I seem to remember trying a list and having an issue, hence why I encoded it in a string. I'll move it to config.

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+        list_of_shapes_to_include = [list_of_shapes_to_include]
+    for geodata_type in list_of_shapes_to_include:
+        shapes = _get_shapes(path_to_shape_dir, geodata_type, scope)
+        all_shapes = all_shapes.append(shapes.geometry)
```
Apparently, newer versions of pandas can concat other types as well, so that `pd.concat([gdf1, gdf2])` returns a GeoDataFrame. Did you try this here to avoid the for loop/append structure?

https://gis.stackexchange.com/questions/162659/joining-concat-list-of-similar-dataframes-in-geopandas

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -128,14 +128,32 @@ rule raw_protected_areas_zipped:
-    shell: "unzip -o {input} -d build/raw-wdpa-feb2019"
+    params:
+        version = config["parameters"]["wdpa-version"]
+    output: temp("build/raw-wdpa")
```
Should this use the `directory` directive?

brynpickering replied:
Do directory and temp work together? Will try.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -128,14 +128,32 @@ rule raw_protected_areas_zipped:
+    message: "Extract protected areas data as zip."
+    input:
+        src = script_dir + "protected_areas_in_europe.py",
+        shapes = rules.raw_protected_areas_in_europe_unzipped.output[0]
```
Rather name this `shape_dir` to make clearer what it is.

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+    all_shapes.to_file(path_to_output, driver=OUTPUT_DRIVER)
+
+
+def _get_shapes(path_to_shapes, geodata_type, scope):
```
`path_to_shapes` -> `path_to_shape_dir `

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+    shapes = gpd.GeoDataFrame()
+    for shapefile in shapefiles:
+        print(f"Reading shape data from {shapefile}")
+        shapes = shapes.append(gpd.read_file(shapefile))
```
Here as well: can we get rid of the append-structure?

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+        print(f"Reading shape data from {shapefile}")
+        shapes = shapes.append(gpd.read_file(shapefile))
+
+    print(f"Number of shapes pre-filter = {len(shapes)}")
```
Can you add "WDPA shapes" to make this clearer in the log?

timtroendle commented on scripts/protected_areas_in_europe.py:
```diff
@@ -0,0 +1,65 @@
+        (shapes.DESIG_ENG != 'UNESCO-MAB Biosphere Reserve')
+    ]
+    shapes = drop_countries(shapes.rename(columns={"ISO3": "country_code"}), scope, print_dropped=False)
+    shapes = drop_geoms_completely_outside_study_area(shapes, scope, print_dropped=False)
```
I understand it's convenient to reuse this function here, but this approach seems too inefficient to me. The WDPA dataset is a beast, and it's global while we need only Europe. A large share of shapes is not needed. I'd go for a more efficient solution.

I guess the simplest is to use a bounding box when reading the shape file using geopandas / fiona.

https://geopandas.org/docs/user_guide/io.html

brynpickering replied:
Dropping countries already handles the issue of problem size. However, I like the idea of using a bounding box on loading in with geopandas. We should do this across all our shapefile loading.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -295,15 +299,8 @@ rule protected_areas_in_europe:
         bounds = "{x_min},{y_min},{x_max},{y_max}".format(**config["scope"]["bounds"])
```
Bounds are unused now.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -268,25 +286,11 @@ rule slope_in_europe:
 
-rule protected_areas_points_to_circles:
-    message: "Estimate shape of protected areas available as points only."
+rule protected_areas_in_europe_rasterised:
```
The naming scheme within this repository was such that `xx_in_europe` indicated a rasterised dataset with same bounds and resolution. I acknowledge this is not a great naming scheme, but I'd stick to it or update it everywhere.

timtroendle commented on rules/data-preprocessing.smk:
```diff
@@ -128,14 +128,32 @@ rule raw_protected_areas_zipped:
+        unzip {input} *.zip -d {output}
+        unzip -o {output}/WDPA_{params.version}_Public_shp_0.zip -d {output}/WDPA_to_merge_0
+        unzip -o {output}/WDPA_{params.version}_Public_shp_1.zip -d {output}/WDPA_to_merge_1
+        unzip -o {output}/WDPA_{params.version}_Public_shp_2.zip -d {output}/WDPA_to_merge_2
```
This approach seems brittle: what if the dataset will contain 4 zips in the future? Can we make this a loop instead?

timtroendle commented on lib/tests/test_shape_utils.py:
```diff
@@ -140,3 +146,47 @@ def test_drop_parts_of_geoms(self, capsys, europe_gdf):
+
+    @pytest.fixture
+    def estimated_polygons(self, points):
+        return estimate_polygons_from_points(points, "REP_AREA")
```
This belongs into the test.

https://docs.pytest.org/en/stable/fixture.html#what-fixtures-are

brynpickering replied:
Hmm, I would say this is part of the arranging process. As given in the example of what fixtures are, the calls to the class being tested are all part of fixtures. This is the same process I'm undertaking here. I then only assert in the test.

timtroendle replied:
I don't see it that way. `estimate_polygons_from_points ` is the behaviour/functionality you are testing, so it belongs into `act` rather than `arrange`.

brynpickering replied:
hmm, ok. I can remove it, but it makes sense as a fixture as it has three tests associated with the same function call. Better to do that once than thrice...

timtroendle replied:
Ok, I checked the pytest doc a little longer and it seems we are both partly right according to their definitions:

> In pytest, “fixtures” are functions you define that serve this purpose. But they don’t have to be limited to just the arrange steps. They can provide the act step, as well, and this can be a powerful technique for designing more complex tests, especially given how pytest’s fixture system works. But we’ll get into that further down.

See this example: https://docs.pytest.org/en/stable/fixture.html#fixtures-can-be-requested-more-than-once-per-test-return-values-are-cached

timtroendle commented on lib/tests/test_shape_utils.py:
```diff
@@ -140,3 +146,47 @@ def test_drop_parts_of_geoms(self, capsys, europe_gdf):
+            assert np.isclose(calculated_radius, radius)
+
+    def test_estimate_polygons_area(self, points, estimated_polygons):
+        assert estimated_polygons.crs == points.crs
```
Usually there's one `assert` per test. This `assert` here seems unnecessary: you are testing this in the test below!

Commit:
Updates following review
